input/listen.go

```diff
@@ -6,42 +6,61 @@ import (
 )
 
 func listen(addr string, handler Handler) error {
-	laddr, err := net.ResolveTCPAddr("tcp", addr)
+	// listeners are set up outside of accept* here so they can interrupt startup
+	listener, err := listenTcp(addr)
 	if err != nil {
 		return err
 	}
-	l, err := net.ListenTCP("tcp", laddr)
+
+	udp_conn, err := listenUdp(addr)
 	if err != nil {
 		return err
 	}
-	log.Notice("listening on %v/tcp", laddr)
-	go acceptTcp(l, handler)
 
-	udp_addr, err := net.ResolveUDPAddr("udp", addr)
+	go acceptTcp(addr, listener, handler)
+	go acceptUdp(addr, udp_conn, handler)
+
+	return nil
+}
+
+func listenTcp(addr string) (*net.TCPListener, error) {
+	laddr, err := net.ResolveTCPAddr("tcp", addr)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	udp_conn, err := net.ListenUDP("udp", udp_addr)
+	listener, err := net.ListenTCP("tcp", laddr)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	log.Notice("listening on %v/udp", udp_addr)
-	go acceptUdp(udp_conn, handler)
-
-	return nil
+	return listener, nil
 }
 
-func acceptTcp(l *net.TCPListener, handler Handler) {
+func acceptTcp(addr string, listener *net.TCPListener, handler Handler) {
+	var err error
+	l := listener
 	for {
-		// wait for a tcp connection
-		c, err := l.AcceptTCP()
+		log.Notice("listening on %v/tcp", addr)
+
+		for {
+			// wait for a tcp connection
+			c, err := l.AcceptTCP()
+			if err != nil {
+				log.Error(err.Error())
+				break
+			}
+			log.Debug("listen.go: tcp connection from %v", c.RemoteAddr())
+			// handle the connection
+			go acceptTcpConn(c, handler)
+		}
+
+		log.Notice("error listening on %v/tcp, reopening connection", addr)
+		l.Close()
+
+		l, err = listenTcp(addr)
```

Comment:
so if the connection is dead we try only 1 time to listen again?  I feel that we should panic if that fails.

Reply:
if we panic (and even if a supervisor restarts) we may lose more data (due to buffers being cleared)
might as well keep the process alive and keep retrying, but maybe with a backoff.

Reply:
ideally, a failure here should propagate up so that eventually main can cleanly shutdown everything, flushing any buffers.  But crng is a long way from having that capability as Main() doesnt even handle TERM signals gracefully.

But retries with a backoff could work, then panic after a configurable amount of time to ensure all buffers would have been cleared.